torch/csrc/api/src/nn/modules/rnn.cpp

```diff
@@ -167,8 +169,8 @@ std::vector<Tensor> RNNImplBase<Derived>::flat_weights() const {
   // (w_ih, w_hh, b_ih, b_hh), repeated for each layer (next to each other).
   std::vector<Tensor> flat;
   const auto num_directions = options.bidirectional_ ? 2 : 1;
-  for (int64_t layer = 0; layer < options.layers_; layer++) {
-    for (auto direction = 0; direction < num_directions; direction++) {
+  for (auto direction = 0; direction < num_directions; direction++) {
+    for (int64_t layer = 0; layer < options.layers_; layer++) {
```

Comment:
From a quick glance at the code, this seems wrong to me. `flat_weights` seems to be used in the same way as the Python counterpart, as [an argument to `_cudnn_rnn_flatten_weight`](https://github.com/pytorch/pytorch/blob/8399197df67f90d7e5cb52d735eb55c803664b35/torch/csrc/api/src/nn/modules/rnn.cpp#L119-L137). In that case, it should have the same order as the parameters in the python `flat_weights`, where the parameters are ordered as [[l0 l0_reversed l1 l1_reversed ...]](https://github.com/pytorch/pytorch/blob/8399197df67f90d7e5cb52d735eb55c803664b35/torch/nn/modules/rnn.py#L65-L86).

Can you add a test for CUDA LSTM forward?

Reply:
I noticed that the tests are only testing for `layer=1` so swapping these for loops shouldn't have an effect

Reply:
I swapped these for loops to make it compatible with [existing RNN.cpp:gather_params() method](https://github.com/pytorch/pytorch/blob/8399197df67f90d7e5cb52d735eb55c803664b35/aten/src/ATen/native/RNN.cpp#L166:L181)

But I will write a test for CUDA LSTM forward to double check

Reply:
Discussed offline, but leaving it here before I forget:
```
In [4]: m = torch.nn.LSTM(input_size=5, hidden_size=7, num_layers=3, bidirectional=True)
In [5]: m._get_flat_weights_names()
Out[5]:
['weight_ih_l0',
 'weight_hh_l0',
 'bias_ih_l0',
 'bias_hh_l0',
 'weight_ih_l0_reverse',
 'weight_hh_l0_reverse',
 'bias_ih_l0_reverse',
 'bias_hh_l0_reverse',
 'weight_ih_l1',
 'weight_hh_l1',
 'bias_ih_l1',
 'bias_hh_l1',
 'weight_ih_l1_reverse',
 'weight_hh_l1_reverse',
 'bias_ih_l1_reverse',
 'bias_hh_l1_reverse',
 'weight_ih_l2',
 'weight_hh_l2',
 'bias_ih_l2',
 'bias_hh_l2',
 'weight_ih_l2_reverse',
 'weight_hh_l2_reverse',
 'bias_ih_l2_reverse',
 'bias_hh_l2_reverse']
```
The python weights appear to be in that order, however, it's possible that they're completely named incorrectly